utilities/fault_injection_fs.cc

```diff
@@ -1093,51 +1201,130 @@ bool FaultInjectionTestFS::TryParseFileName(const std::string& file_name,
   return ParseFileName(file, number, type);
 }
 
-IOStatus FaultInjectionTestFS::InjectWriteError(const std::string& file_name) {
-  MutexLock l(&mutex_);
-  if (!enable_write_error_injection_ || !write_error_one_in_) {
+IOStatus FaultInjectionTestFS::MaybeInjectThreadLocalWriteError(
+    const IOOptions& io_options) {
+  ErrorContext* ctx =
+      static_cast<ErrorContext*>(injected_thread_local_write_error_->Get());
+  if (ctx == nullptr || !ctx->enable_error_injection || !ctx->one_in ||
+      io_activties_exempted_from_fault_injection.find(io_options.io_activity) !=
+          io_activties_exempted_from_fault_injection.end()) {
     return IOStatus::OK();
   }
-  bool allowed_type = false;
 
-  if (inject_for_all_file_types_) {
-    allowed_type = true;
-  } else {
-    uint64_t number;
-    FileType cur_type = kTempFile;
-    if (TryParseFileName(file_name, &number, &cur_type)) {
-      for (const auto& type : write_error_allowed_types_) {
-        if (cur_type == type) {
-          allowed_type = true;
-        }
-      }
+  IOStatus ret;
+  if (ctx->rand.OneIn(ctx->one_in)) {
+    ctx->count++;
+    if (ctx->callstack) {
+      free(ctx->callstack);
     }
+    ctx->callstack = port::SaveStack(&ctx->frames);
+    ctx->message = "injected write error";
+    ret = IOStatus::IOError(ctx->message);
+    ret.SetRetryable(ctx->retryable);
+    ret.SetDataLoss(ctx->has_data_loss);
+  }
+  return ret;
+}
+
+IOStatus FaultInjectionTestFS::MaybeInjectThreadLocalMetadataReadError(
```

Comment:
More copy-paste that is a pain for maintenance. (There are a couple ways to handle SYNC_POINT without triggering strlen on each call. Ping me if you want advice.)

Reply:
Fixed